packages/@aws-cdk/cdk/lib/cross-environment-token.ts

```diff
@@ -0,0 +1,66 @@
+import { ArnComponents } from './arn';
+import { DeployTimeOrAssignedPhysicalName } from "./deploy-time-or-assigned-physical-name";
+import { PhysicalName } from './physical-name';
+import { IResource } from './resource';
+import { ResolveContext, Token } from './token';
+
+/**
+ * A Token that represents a reference that spans accounts and/or regions,
+ * and so requires the resources to have physical names.
+ * You should never need to interact with these directly,
+ * instead use the {@link ResourceIdentifiers} class.
+ * This class is private to the @aws-cdk/cdk package.
+ */
+export abstract class CrossEnvironmentToken extends Token {
+  private readonly resource: IResource;
+  private readonly physicalName: PhysicalName;
+
+  /**
+   * @param regularValue the value used when this is referenced NOT from a cross account and/or region Stack
+   * @param crossEnvironmentValue the value used when this is referenced from a cross account and/or region Stack
+   * @param resource the scope this reference is mastered in. Used to determine the owning Stack
+   * @param physicalName
+   * @param displayName a short name to be used in Token display
+   */
+  protected constructor(private readonly regularValue: string, private readonly crossEnvironmentValue: any,
+                        resource: IResource, physicalName: PhysicalName, displayName: string) {
+    super(undefined, displayName);
+
+    this.resource = resource;
+    this.physicalName = physicalName;
+  }
+
+  public resolve(context: ResolveContext): any {
+    const consumingStack = context.scope.node.stack;
+    const owningStack = this.resource.node.stack;
+
+    if (consumingStack.env.account !== owningStack.env.account ||
+        consumingStack.env.region !== owningStack.env.region) {
+      // check whether the resource has a physical name set
+      if (!this.physicalName.availableAtSynthesisTime) {
```

Comment:
Separation of concerns could be improved. I feel like instead of exposing `availableAtSynthTime` and `_setName` you could just expose something like `physicalName.resolveNow(this.resource)` which will have all this logic of determining if this physical name can be resolved now and if it can, it will return it or generate it and return it at it's own discretion.

Reply:
Excellent point. After doing what you suggested, I took the refactoring even further, and moved `physicalName` to `(I)Resource`, and so it was not needed anymore in `CrossEnvironmentToken` (it felt off to me that we were passing both `resource` and `physicalName` there, now it looks much cleaner to me). Let me know what you think of this new version!